### PR TITLE
Diff as update

### PR DESCRIPTION
**Pull request by langsamu**
# Summary
Adds an extension method to [`GraphDiffReport`](https://github.com/dotnetrdf/dotnetrdf/blob/5b7fc480346c90eb9b164fb4f8ee09f378442d52/Libraries/dotNetRDF/Core/GraphDiff.cs#L321) that translates it into an equivalent [`ModifyCommand`](https://github.com/dotnetrdf/dotnetrdf/blob/5b7fc480346c90eb9b164fb4f8ee09f378442d52/Libraries/dotNetRDF/Update/Commands/ModifyCommand.cs#L43).

# Logic
- Removed ground triples are added as is to both delete and where clauses.
- Blank nodes in removed non-ground triples are converted to variables and added to both delete and where clauses.
- An [`isBlank`](https://www.w3.org/TR/sparql11-query/#func-isBlank) filter is added for each blank node in removed non-ground triples.
- Added triples (ground or not) are added as is to the insert clause.

# Example
## Old graph
s | p | o
-- | -- | --
`urn:WontDeleteOrInsert` | `urn:p` | `urn:o`
`urn:WillDelete` | `urn:p` | `urn:o`
`_:WillConvertFilterDelete` | `urn:p1` | `urn:o`

## New graph
s | p | o
-- | -- | --
`urn:WontDeleteOrInsert` | `urn:p` | `urn:o`
`urn:WillInsert` | `urn:p` | `urn:o`
`_:WillInsert` | `urn:p2` | `urn:o`

## Generated SPARQL
```SPARQL
DELETE
{ 
  <urn:WillDelete> <urn:p> <urn:o> . 
  ?WillConvertFilterDelete <urn:p1> <urn:o> . 
}
INSERT
{ 
  _:WillInsert <urn:p2> <urn:o> . 
  <urn:WillInsert> <urn:p> <urn:o> . 
}
WHERE
{ 
  <urn:WillDelete> <urn:p> <urn:o> . 
  ?WillConvertFilterDelete <urn:p1> <urn:o> . 
  FILTER(ISBLANK(?WillConvertFilterDelete)) 
}
```

### CODE REVIEW

**Comment by tpluscode** [comment]
Thanks @langsamu, that's actually kinda cool :)

Before I even look closely at the code, do you think there are any edge cases which you may want to add as test cases? 

I also find it interesting how you used the query builder's internals. Guess it's high time it we added the `UPDATE`/`MODIFY` query forms too :)

**Comment by langsamu** [comment]
Thanks, we enjoyed it. Credit @christopheralcock for the actual algorithm.

An actual update builder was my initial direction, I was surprised to realise there isn't one. High time for sure. I'm afraid I'm not comfortable enough with this to tackle.

Didn't find edge cases other than filtering for blanks. All this does is dispatch between diff and command, and the logic exhausts (as far as I can tell) the capabilities of the diff.

A serious caveat is that the generated SPARQL is named graph agnostic, which of course is naive. But the diff is agnostic as well, so there's not much I want to do with it.

The philosophy is that we build a `ModifyCommand`, intentionally not a string, so callers can customise it further (our initial use-case is updating the ontology in GraphDB, which requires a _schemaTransaction_ marker statement in the INSERT clause).
Support for named graphs and any other SPARQL I couldn't get from the diff could also be added by the caller.

Happy to enhance if someone can help identify more edge cases.

**Comment by tpluscode** [comment]
Good point about named graphs. Would is simply mean wrapping patterns inside `DELETE`/`INSERT`/`WHERE` in `GRAPH {}`? 

Maybe it could be an overload taking an additional parameter?

```c#
public static ModifyCommand AsUpdate( 
    this GraphDiffReport diff,
    Uri graph,
    INamespaceMapper prefixes = null)
{
}
```

Implementation would not change much, just extract a private method taking the three builders and in the overload, calling `builder.Graph()` to wrap the patterns.

**Comment by langsamu** [comment]
I've added support for the [`WITH` clause](https://www.w3.org/TR/sparql11-update/#deleteInsert), which does exactly what you recommended. `ModifyCommand` handles that with a `graphUri` in the [constructor](https://github.com/dotnetrdf/dotnetrdf/blob/5b7fc480346c90eb9b164fb4f8ee09f378442d52/Libraries/dotNetRDF/Update/Commands/ModifyCommand.cs#L55).

Also added tests to highlight what happens when either graph in the diff is null (all inserts, all deletes or nothing).

File: Libraries/dotNetRDF/Core/GraphDiffReportExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+        private static void AddBlankNodeFilter(this IGraphPatternBuilder builder, IBlankNode n)
+        {
+            builder.Filter(node =>
+                node.IsBlank(n.InternalID));
+        }
+
+        private static INode AsVariable(this INode n)
+        {
+            return n is IBlankNode blank ? new NodeFactory().CreateVariableNode(blank.InternalID) : n;
```
[the remark this answers is not in the document]
Comment (kal) [reply Libraries/dotNetRDF/Core/GraphDiffReportExtensions.cs]: I'd agree with @tpluscode. Although you aren't using the blank node mapper in the NodeFactory instance right now, it would be good to be using a single factory for the whole of the command generation process just to avoid a bunch of short term objects on the heap.
